### PR TITLE
New input

### PR DESCRIPTION
**Pull request by jobordner**
[ edit: removed bug warning, fixed in 0ef60671e ]

This branch implements scalable checkpoint / restart, and allows for changing parameters and numbers of processors on restart. It's another huge PR, sorry about that.

Example test files are in `input/Checkpoint/test_cosmo-[check|restart].in`, which run the DD cosmology test as usual to cycle 160, but writes a checkpoint directory at cycle 140 and restarts from that. 

This is scalable: it works by generating a Morton curve ordering of blocks, and assigning blocks to `IoWriter` (for checkpoint) or `IoReader` (for restart) chare array elements, which handle concurrent I/O. No parallel HDF5 (yet), but checkpoint has been tested on 212992 cores writing 4096 files.

There are some remaining issues to clean up, hence the "draft pull request". Including but not necessarily limited to:

1. finishing documentation
2. fixing some memory leaks
3. more thorough testing
4. integrating the `"order"` method into the `"check"` method
5. cleaning up code a bit

### CODE REVIEW

**Comment by peeples** [comment]
@jobordner What is the current status of this, as it is one of the current highest priority items? In particular, would it be possible for people to start reviewing it now, with some of the other items (e.g., documentation, the code cleanup) as part of another, smaller, later PR?

**Comment by jobordner** [comment]
> 

@peeples Sure, it's usable though there are some remaining code issues (e.g. memory leak).

**Review by jwise77 — CHANGES_REQUESTED** [review]
Because this was out of my wheelhouse, I could only do a superficial review of the code. I made several suggestions and questions in the code -- no development blockers, just some clarification and additional documentation.

File: input/test_cosmo.incl [hunk 1]
```diff
@@ -307,5 +308,5 @@ Adapt {
  }
 
  Stopping {
-     cycle = 100;
+     redshift = -50.0;
```
[the remark this answers is not in the document]
Comment (jobordner) [reply input/test_cosmo.incl]: This was a temporary change for testing something out that got inadvertently checked in. Reverted.

**Review by mabruzzo — APPROVED** [review]
I took a quick look through all of the files. I left a few small comments, but it generally looks good to me. The documentation looks like it needs a little more work. But, maybe it makes sense to merge sooner?

At some point we really need to document all of the output parameters on the webpage. I suppose it will make sense to do this for `EnzoMethodOutput` and `EnzoMethodCheck`, when we finally rip off the band-aid and remove the old output method.

Do I understand correctly that `EnzoMethodOutput` and `EnzoMethodCheck` use different strategies for assigning blocks to data files? If so, is the long-term intention to unify these strategies or keep them different?

File: src/Enzo/enzo_control_restart.cpp [hunk 2]
```diff
@@ -0,0 +1,874 @@
...
+    } else if (type_meta == "file") {
+      file->file_read_meta(buffer,name.c_str(),&type_scalar,&nx,&ny,&nz);
+    } else {
+      ERROR1 ("MethodOutput::read_meta_()",
+              "Unknown type_meta \"%s\"",
+              type_meta.c_str());
+    }
+    // Get object's ith metadata
+
+    io->meta_value(i,& buffer, &name, &type_scalar, &nx,&ny,&nz);
```
Comment (mabruzzo) [inline src/Enzo/enzo_control_restart.cpp:815]:
```suggestion
```
I think you can delete these lines

**Comment by jobordner** [comment]
> Do I understand correctly that `EnzoMethodOutput` and `EnzoMethodCheck` use different strategies for assigning blocks to data files? If so, is the long-term intention to unify these strategies or keep them different?

The intention is to unify them, such that `EnzoMethodOutput` uses the `EnzoMethodCheck` approach.

File: src/Enzo/enzo_control_restart.cpp [hunk 3]
```diff
@@ -0,0 +1,874 @@
...
+  contribute(callback);
+}
+
+//----------------------------------------------------------------------
+
+void Simulation::r_restart_start (CkReductionMsg * msg)
+{
+  TRACE_SIMULATION("EnzoSimulation::r_restart_start()",this);
+  delete msg;
+  // [ Called on root ip only ]
```
[the remark this answers is not in the document]
Comment (jobordner) [reply src/Enzo/enzo_control_restart.cpp:175]: my convention for variable storing process id; will clarify

File: src/Enzo/enzo_control_restart.cpp [hunk 4]
```diff
@@ -0,0 +1,874 @@
...
+
+    } else {
+
+      // // Block doesn't exist: create it and send its data
+      // enzo::factory()->create_block_check
+      //   ( msg_check, enzo::block_array(),index );
+
+      // // Also tell parent that it has children (surprise!)
+      // Index index_parent = index.index_parent(cello::config()->mesh_min_level);
+    }
```
[the remark this answers is not in the document]
Comment (jobordner) [reply src/Enzo/enzo_control_restart.cpp:301]: Will remove, this case is not needed.

File: src/Cello/io_IoFieldData.cpp [hunk 5]
```diff
@@ -115,13 +115,12 @@ void IoFieldData::field_array
...
+
 //----------------------------------------------------------------------
 
-void IoFieldData::particle_array 
-(int it, int ib, int ia,
- void ** buffer, std::string * name, int * type,
- int * n, int * k) throw()
+void IoFieldData::save_to (void * v)
 {
+  INCOMPLETE("IoFieldData::save_to()");
```
[the remark this answers is not in the document]
Comment (jobordner) [reply src/Cello/io_IoFieldData.cpp:123]: This is not needed--removed.

File: src/Enzo/enzo-e.cpp [hunk 6]
```diff
@@ -71,23 +73,28 @@ PARALLEL_MAIN_BEGIN
 #endif
   
 
-  // Check parameter file
+  // Check and parse arguments
 
-  if (PARALLEL_ARGC != 2) {
+  bool dryrun = false;
+  if (PARALLEL_ARGC > 2 && strcmp(PARALLEL_ARGV[2],"-dryrun")==0) {
```
[the remark this answers is not in the document]
Comment (jobordner) [reply src/Enzo/enzo-e.cpp:79]: added usage printf

File: doc/source/design/design-input.rst [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+   s0 -> r1
+   destroy r1
+   s0 ->o b0 : p_restart_done()
+   s0 ->o bk
+   s0 ->o bkp1
+
+   @enduml
+
+// Begin reading restart data and create the mesh hierarchy of
+// EnzoBlocks. Replaces functionality of control_adapt.
```
[the remark this answers is not in the document]
Comment (jobordner) [reply doc/source/design/design-input.rst]: Agreed; done.

**Review by jwise77 — APPROVED** [review]
Thanks for addressing my comments. I'm approving, but before merging, there are some conflicts that need to be fixed.

**Comment by peeples** [comment]
@jobordner looks like this is ready to go, modulo the conflicts (from @brittonsmith's new answer tests?). Can we get this through in time for a July 1 v1.0 release?

**Pull request merged**